### Why there are so many `None` in the `songplays` table?

In [3]:
import os
import glob
import pandas as pd

In [4]:
# Define a function return concatenate all the data into one dataframe.

def process_file(filepath):
    all_files = []
    for root, dirs, files in os.walk(filepath):
        files = glob.glob(os.path.join(root,'*.json'))
        for f in files :
            all_files.append(os.path.abspath(f))

    # get total number of files found
    num_files = len(all_files)
    print('{} files found in {}'.format(num_files, filepath))
    file=all_files[0]
    df_new = pd.read_json(file,lines=True)
    columns=df_new.columns
    
    df = pd.DataFrame(columns=columns)
    for file in all_files:
        df_new = pd.read_json(file,lines=True)
        df = pd.concat([df,df_new],ignore_index=True)
    return df


In [7]:
df_song =process_file('data/song_data') # store all the song_data into one dataframe

83 files found in data/song_data


In [8]:
df_song.head()

artist_id  artist_latitude              artist_location  \
0  AR7G5I41187FB4CE6C              NaN              London, England   
1  AR8ZCNI1187B9A069B              NaN                                
2  ARXR32B1187FB57099              NaN                                
3  AR10USD1187B99F3F1              NaN  Burlington, Ontario, Canada   
4  ARGSJW91187B9B1D6B         35.21962               North Carolina   

   artist_longitude            artist_name   duration num_songs  \
0               NaN               Adam Ant  233.40363         1   
1               NaN       Planet P Project  269.81832         1   
2               NaN                    Gob  209.60608         1   
3               NaN  Tweeterfriendly Music  189.57016         1   
4         -80.01955           JennyAnyKind  218.77506         1   

              song_id            title  year  
0  SONHOTT12A8C13493C  Something Girls  1982  
1  SOIAZJW12AB01853F1       Pink World  1984  
2  SOFSOCN12A8C143F5D   Face the Ashes  2007  
3  SOHKNRJ12A6701D1F8     Drop of Rain     0  
4  SOQHXMF12AB0182363  Young Boy Blues     0

In [9]:
df_log =process_file('data/log_data') # store all the log_data into one dataframe

30 files found in data/log_data


In [10]:
df_log.head()

artist       auth firstName gender itemInSession  \
0                   Frumpies  Logged In  Anabelle      F             0   
1  Kenny G with Peabo Bryson  Logged In  Anabelle      F             1   
2                Biffy Clyro  Logged In  Anabelle      F             2   
3                       None  Logged In      Lily      F             0   
4                        HIM  Logged In      Lily      F             1   

  lastName     length level                                     location  \
0  Simpson  134.47791  free  Philadelphia-Camden-Wilmington, PA-NJ-DE-MD   
1  Simpson  264.75057  free  Philadelphia-Camden-Wilmington, PA-NJ-DE-MD   
2  Simpson  189.83138  free  Philadelphia-Camden-Wilmington, PA-NJ-DE-MD   
3    Burns        NaN  free        New York-Newark-Jersey City, NY-NJ-PA   
4    Burns  212.06159  free        New York-Newark-Jersey City, NY-NJ-PA   

  method      page  registration sessionId                            song  \
0    PUT  NextSong  1.541044e+12       455                      Fuck Kitty   
1    PUT  NextSong  1.541044e+12       455  By The Time This Night Is Over   
2    PUT  NextSong  1.541044e+12       455                     God & Satan   
3    GET      Home  1.540621e+12       456                            None   
4    PUT  NextSong  1.540621e+12       456                       Beautiful   

  status             ts                                          userAgent  \
0    200  1541903636796  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4...   
1    200  1541903770796  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4...   
2    200  1541904034796  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4...   
3    200  1541910841796  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...   
4    200  1541910973796  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...   

  userId  
0     69  
1     69  
2     69  
3     32  
4     32

In [11]:
print(df_song.shape, df_log.shape)

(83, 10) (8056, 18)


In [13]:
song_set = set(df_song['title'])
song_log_set = set(df_log['song'])
inter = song_set.intersection(song_log_set)
print("df_song contains {} songs.".format(len(song_set)))
print("df_log contains {} songs.".format(len(song_log_set)))
print("Only {} songs in both of the two dataframes.".format(len(inter)))

df_song contains 71 songs.
df_log contains 5190 songs.
Only 2 songs in both of the two dataframes.


### Conclusion
Sparkify provides only two datasets:
* song dataset, which contains 71 different songs,
* log dataset, which contains 5190 different songs.

Therefore, for the songs in log dataset, we can not find their information, like `song_id` and `artist_id` which are stored in the song dataset. Apparently, Sparkify does not provid enough data for the song data. 